# Meningioma_DNA_methylation_project
Our study highlights the role of a DNA hypermethylation signature and its downstream effects on signaling pathways involving clustered protocadherin genes during progression of meningiomas.

This study includes DNA methylation array data from a cross-sectional meningioma cohort (n=231), and a longitudinal cohort including primary and recurrent meningioma samples from a total of 18 patients.

All meningiomas were profiled using the MethylationEPIC v2.0 array. Raw data are archived at GEO, and meta data is available from the corresponding publication listed below. This repository provides essential codes used: 

I) to identify a DNA hypermethylation signature associated with disease progression in meninigioma
II) to infer copy number variations in benign and malignant meningiomas
III) to perform survival analyses and assess prediction performances of Cox regression models
IV) to investigate differential methylation in between normal meninges and distinct meningioma settings
V) to perform trajectory inference in order to analyze DNA methylation profiles in the context of microevolutionary development from normal meninges to malignant meningioma
VI) to perform differential gene expression analysis of benign and malignant meningiomas

Lastly, we provide an easy-to-use web application that predicts hypermethylation class association and genome-wide copy number analysis at https://mmcc.neurologie.uni-tuebingen.de.

